### ISSUE TITLE
Enlarge region of user mouse input more than monitor

### ISSUE THREAD

**Issue by JaroslavHolecek** [ocornut/imgui#4440]
Hi,
I am trying to do scaleable content. I know https://github.com/thedmd/imgui-node-editor and I am trying to do something similar.
I am stuck on how to capture mouse events at positions outside of monitor.

I do it like this (and problem is with zoom-out -> making things smaller ):
1) Enlarge ClipRects of window and window ImDrawList (the same for all parent windows)
2) Transform mouse to opposite-scaled-coords (because in result view are for example DragFloat elements smaller than normal, so small mouseDelta in ImGui is large mouseDelta for my elements )
3) Draw everything I need
4) Scale all vertices in ImDrawList->VtxBuffer (make elements smaller)

My actual problem is in step 1) and 2):
Because I enlarge area of ClipRects and region where user (virtually) manipulate with mouse. I can go with mouse outside of monitor. And if so, ImGui not process events (hoovering elements, clicking) on elements (that are virtually outside of monitor too).

What I have to enlarge moreover for allow ImGui to process events outside of monitor?

**Comment by ocornut** [comment]
User interactions within a window are affected by the current active window's ClipRect, which is initially clipped with the host viewport rectangle.

```
const ImRect host_rect = viewport_rect;
const ImRect outer_rect = window->Rect();
window->OuterRectClipped = outer_rect;
window->OuterRectClipped.ClipWith(host_rect);
[...]
PushClipRect(host_rect.Min, host_rect.Max, false);
```

Disabling the clipping would affect a large amount of things in dear imgui so I wouldn't recommend it.

Instead you should rework your scaling strategy to make it that the time of widgets submission, the window active ClipRect is scaled and modified accordingly.

**Comment by JaroslavHolecek** [comment]
OK, I rework it in way of enlarge/shrink font size and objects (in ImDrawList) generated by my app -> so I submit elements in already zoomed size.
Are here some better option? -> include all elements (and objects to ImDrawList) and than at the end zoom(out) everything - with correct user interaction (on smaller/larger elements)?

Because now I have to compute (zoomed) size of every element individually ( in different place of code ) with increase possibility of bug.
Thank you.

**Comment by ocornut** [comment]
As per my answer:

> Instead you should rework your scaling strategy to make it that the time of widgets submission, the window active ClipRect is scaled and modified accordingly.

Doesn’t mean the drawlist contents other than ClipRect need to be transformed ahead. You can still late transform if that works for you but need to make sure the high level ImGui side clip rect contains the post-transform area you are working with.

**Comment by JaroslavHolecek** [comment]
OK - and that is probably my problem (maybe misunderstanding).
If I want to "zoom out" everything at the end (so in time of including elements are all of them larger, same for whole workarea - basically I draw everything large (so whole workarea is large) and than scale it to size of window). I submit on places that can be outside of window/application/monitor - and after this submitions I move/scale all of elements to be inside of actual window . And in this place (outside of window/app/monitor) user interaction do not work.
Maybe I just miss some trivial principle of ClipRects...

**Comment by rokups** [comment]
Why would you need interaction with invisible UI elements to begin with?

**Comment by JaroslavHolecek** [comment]
> Why would you need interaction with invisible UI elements to begin with?

Because after scaling them (zooming) they are not invisible.

**Comment by rokups** [comment]
Yes, i understand what you are developing, i even implemented [same thing](https://github.com/rokups/ImNodes/blob/master/ImNodes.cpp) myself. I am still confused why you need interaction with invisible items. Sure, after scaling they are invisible. You can not click what you do not see though. CTRL+F in linked file for "->Zoom" for some inspiration. Scaling is nothing special, all it does is adjusting positions and sizes of elements, nothing else. No issues with clipping or anything, because as far as Dear ImGui is concerned - there is no scaling going on, just some things are rendered using a specific size and thats it.

**Comment by JaroslavHolecek** [comment]
> Yes, i understand what you are developing, i even implemented [same thing](https://github.com/rokups/ImNodes/blob/master/ImNodes.cpp) myself. I am still confused why you need interaction with invisible items. Sure, after scaling they are invisible. You can not click what you do not see though. CTRL+F in linked file for "->Zoom" for some inspiration. Scaling is nothing special, all it does is adjusting positions and sizes of elements, nothing else. No issues with clipping or anything, because as far as Dear ImGui is concerned - there is no scaling going on, just some things are rendered using a specific size and thats it.

I try to inspect your file. Maybe (probably) I just do not understand some ImGui core funkcionality. I try to explain it for example on DragFloat.
In time of including DragFloat it is on 100,100-150,200 rectangle and here ImGui catch interaction with it. If I scale vertexes by 0.5 in ImDrawList, DragFloat is drawn on 50,50-125,100 rectangle, but interaction are still on 100,100-150,200 rectangle. I try to solve this by leave mouse position on original (large) scale. That work fine. But if something scaled (small) is in bottom of window - it is outside of window/app/monitor in original coordinates. If my monitor has width 100 pixels you can see it in example values with DragFlow

**Comment by rokups** [comment]
Ah i see, you do scaling by directly manipulating vertex buffer. That explains it. It is an interesting approach. My way of doing things is more complicated and is not perfect. One thing that comes to mind is that you should have no issues when you zoom in. You are only having issues when you zoom out and start displaying UI elements that are technically outside of view and get clipped by default, right? If that is the case you could try rendering your UI elements at max zoomed out level so everything is tiny and is visible on the screen, then you should have no issues with clipping. Then scale vertex buffers to zoom in and achieve normal-looking UI size. Clipping rects could be also scaled down to reintroduce clipping when zoomed in. Should you succeed - please let us know. Maybe i should adopt this approach myself 👍🏻

**Comment by JaroslavHolecek** [comment]
> Ah i see, you do scaling by directly manipulating vertex buffer. That explains it. It is an interesting approach. My way of doing things is more complicated and is not perfect. One thing that comes to mind is that you should have no issues when you zoom in. You are only having issues when you zoom out and start displaying UI elements that are technically outside of view and get clipped by default, right? If that is the case you could try rendering your UI elements at max zoomed out level so everything is tiny and is visible on the screen, then you should have no issues with clipping. Then scale vertex buffers to zoom in and achieve normal-looking UI size. Clipping rects could be also scaled down to reintroduce clipping when zoomed in. Should you succeed - please let us know. Maybe i should adopt this approach myself 👍🏻

I adopt method from linked node-editor, but can not figure out how to deal with this (can not find it node-editor code)
Yes, you are correct in (as far as I know) all points. Issue happens only when doing elements smaller. Drawing all small is impossible (or very ineffective) since canvas can be "infinite" and there will be a lots of elements even on monitor in smallest scale and most of time will be user in much larger scale.
Scale clipping rects is what I am trying to do, but I am stuck on clipping rect of monitor.
